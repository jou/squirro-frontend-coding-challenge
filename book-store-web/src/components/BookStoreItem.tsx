import styled from "styled-components";
import { BookStoreAttributes } from "~/api/Client";
import { HTMLAttributes } from "react";
import { Avatar } from "~/components/Avatar";

const RATINGS = [1, 2, 3, 4, 5];

const BookStoreItemWrapper = styled.div`
    display: flex;
    flex-direction: column;

    border: 1px gray solid;
    border-radius: 0.5rem;
    padding: 1rem;

    .book-store-item__main {
        display: flex;
    }

    .book-store-item__store-image {
        flex: 0 0 8rem;
        margin: 0 1rem 1rem 0;
    }

    .book-store-item__info {
        flex: 1;
        display: flex;
        flex-direction: column;
    }

    .book-store-item__info-top {
        display: flex;
        justify-content: space-between;
        align-items: center;
    }

    .book-store-item__title {
        margin: 0;
    }

    .book-store-item__rating {
        font-size: 1.5rem;
    }

    .book-store-item__meta {
    }
`;

export interface BookStoreItemProps {
    bookStore: BookStoreAttributes;
}

export default function BookStoreItem({
    bookStore,
    ...wrapperProps
}: BookStoreItemProps & HTMLAttributes<HTMLDivElement>): JSX.Element {
    const ratingStars: string = RATINGS.map((rating) => {
        if (bookStore.rating >= rating) {
            return "★";
        }

        return "☆";
    }).join("");

    return (
        <BookStoreItemWrapper
            {...wrapperProps}
            className={`book-store-item ${wrapperProps.className}`}
        >
            <div className="book-store-item__main">
                <Avatar
                    alt={`Picture of ${bookStore.name}`}
                    className="book-store-item__store-image"
                    avatarUrl={bookStore.storeImage}
                />
                <div className="book-store-item__info">
                    <div className="book-store-item__info-top">
                        <h1 className="book-store-item__title">
                            {bookStore.name}
                        </h1>
                        <span className="book-store-item__rating">
                            {ratingStars}
                        </span>
                    </div>
                </div>
            </div>
            <div className="book-store-item__meta">
                {bookStore.establishmentDate}
            </div>
        </BookStoreItemWrapper>
    );
}
