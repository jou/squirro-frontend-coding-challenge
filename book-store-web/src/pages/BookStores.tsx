import { useApiClient, useBookStores } from "~/hooks/useApiClient";
import BookStoreList from "~/components/BookStoreList";

export default function BookStores(): JSX.Element {
    const apiClient = useApiClient();
    const { loading, error, data: bookStores } = useBookStores(apiClient);

    if (loading) {
        return <div>loading...</div>;
    }
    if (error) {
        return <div>{error.message}</div>;
    }
    return <BookStoreList bookStores={bookStores?.data ?? []} />;
}
