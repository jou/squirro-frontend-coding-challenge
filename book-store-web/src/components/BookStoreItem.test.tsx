import { describe, expect, it } from "vitest";
import { render } from "@testing-library/react";
import BookStoreItem from "~/components/BookStoreItem";
import { AwesomeBookStore } from "~/tests/fixtures/BookStores";

describe("<BookStoreItem>", () => {
    // Not much interesting happening here so far, a snapshot is enough for now
    it("should match snapshot", () => {
        const { container } = render(
            <BookStoreItem bookStore={AwesomeBookStore} />,
        );

        expect(container).toMatchSnapshot();
    });
});
