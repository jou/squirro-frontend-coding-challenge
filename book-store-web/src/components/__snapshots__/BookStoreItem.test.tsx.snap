// Vitest Snapshot v1

exports[`<BookStoreItem> > should match snapshot 1`] = `
<div>
  <div
    class="sc-hKMtZM hVrONt book-store-item undefined"
  >
    <div
      class="book-store-item__main"
    >
      <div
        class="sc-bczRLJ gHExxr avatar book-store-item__store-image"
      >
        <img
          alt="Picture of Awesome Book Store"
          class="avatar__image"
          src="http://example.com/bookstore/store.jpg"
        />
      </div>
      <div
        class="book-store-item__info"
      >
        <h1
          class="book-store-item__title"
        >
          Awesome Book Store
        </h1>
        <span
          class="sc-dkzDqf ljtOeY star-rating book-store-item__rating"
        >
          <span
            class="star-rating__star"
            data-testid="star-rating__star-1"
            role="button"
          >
            â˜…
          </span>
          <span
            class="star-rating__star"
            data-testid="star-rating__star-2"
            role="button"
          >
            â˜…
          </span>
          <span
            class="star-rating__star"
            data-testid="star-rating__star-3"
            role="button"
          >
            â˜…
          </span>
          <span
            class="star-rating__star"
            data-testid="star-rating__star-4"
            role="button"
          >
            â˜†
          </span>
          <span
            class="star-rating__star"
            data-testid="star-rating__star-5"
            role="button"
          >
            â˜†
          </span>
        </span>
      </div>
      <table
        class="sc-gsnTZi bodUZA bestseller-table book-store-item__bestseller"
      >
        <colgroup>
          <col
            class="bestseller-table__title-col"
          />
          <col
            class="bestseller-table__author-col"
          />
        </colgroup>
        <thead>
          <tr>
            <th
              colspan="2"
            >
              Best-selling books
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            data-testid="bestseller-table__row"
          >
            <td>
              Awesomest Book
            </td>
            <td>
              Unknown Author
            </td>
          </tr>
          <tr
            data-testid="bestseller-table__row"
          >
            <td>
              Awesomer Book
            </td>
            <td>
              Johner Doe
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="book-store-item__meta"
    >
      <span>
        <span>
          01.01.2020
        </span>
         -
         
        <a
          href="http://example.com/bookstore"
        >
          example.com
        </a>
      </span>
      <span
        class="book-store-item__flag"
      >
        ðŸ‡¨ðŸ‡­
      </span>
    </div>
  </div>
</div>
`;
